### ISSUE TITLE
review `std/crypto` and `std/hash`

### ISSUE THREAD

**Issue by iuioiua** [denoland/deno_std#2733]
Continuing the discussion from #2719, `std/hash` shouldn't be fully deprecated due to its desired streaming functionality that isn't yet available as a web spec. [Although, it's planned to one day be.](https://github.com/wintercg/proposal-webcrypto-streams)

However, `std/hash` does overlap with `std/crypto` in a few ways. I have a few suggestions that I think would prune some of this overlapped functionality:
1. Move stream-based functionality within `std/hash` to `std/crypto/streams`.
2. Deprecate APIs within `std/hash` that exist within `std/crypto`.
3. Move some `Hasher` methods such as `toString` as pure functions in `std/crypto`.

Put simply; I suggest we merge `std/hash` into `std/crypto`. Doing so would get rid of duplicate code and maintain streaming support. @kt3k and @cjihrig, what do you guys think?

**Comment by cjihrig** [comment]
If we can make `std/hash` completely redundant, I'd be supportive of moving things into `std/crypto`. If we can't make it redundant, maybe we should just wait on streaming webcrypto.

**Comment by iuioiua** [comment]
@kt3k, would you be able to point out where streaming functionality exists within `std/hash`. Having a quick look through, I'm unable to spot it. Maybe I'm missing something...

**Comment by iuioiua** [comment]
I think `std/crypto` would benefit from having a `toHashString` function, inspired by the `Hash.toString()` method from `std/hash`. Seems like a convenient thing to have.

Apart from that, I don't see where `std/crypto` is lacking, compared to `std/hash` (see above comment).

**Comment by iuioiua** [comment]
Perhaps, there may be some confusion between `std/hash` and `node:crypto` (which does support streams with hashes).

**Comment by kt3k** [comment]
@iuioiua 
With `std/hash`, you can call `hash.update("...")` multiple times and that appends the data to the internal state.

```js
const hash = createHash("md5");
for await (const chunk of readable) {
  hash.update(chunk);
}
hash.digest().toString();
```

I think we can consider this capability as streaming support. (The same thing is not possible with current Web Crypto's `digest` API, I guess)

**Comment by iuioiua** [comment]
Ok, I see what you mean now. Well, I think `std/crypto` can be easily tweaked to support streams. By adding `ReadableStream<Uint8Array>` as a possible argument type for `crypto.subtle.digest()`, I was able to get the following working (taken from [proposal explainer](https://github.com/wintercg/proposal-webcrypto-streams/blob/main/explainer.md)):
```ts
import { crypto } from "./crypto/mod.ts";
import { assertEquals } from "./testing/asserts.ts";

const algorithm = "SHA-256";

// New: digest supports streaming
const response1 = await fetch("https://http.cat/200");
const digest1 = await crypto.subtle.digest(algorithm, response1.body!);

const response2 = await fetch("https://http.cat/200");
const digest2 = await crypto.subtle.digest(
  algorithm,
  await response2.arrayBuffer(),
);

assertEquals(digest1, digest2);
```
There are also `EncryptionStream` and `DecryptionStream` from the same proposal which might be able to be polyfilled. I'm no expert on streams or crypto, but I can give this a crack (with some help).

**Comment by iuioiua** [comment]
@kt3k, I'll go ahead and remove `std/hash` if and once the above PRs are merged, seeing as it's been deprecated for a while now.
